### DISCUSSION TITLE
Upgrading to v5 and type-safe populate

### DISCUSSION THREAD

**Discussion by pphunor72** [mikro-orm/mikro-orm#2812] [Q&A]
Hello,

I cannot figure out how to set the type of find options properly if it has a populate property. Calling directly the findOne works fine:

		const user = await this.userRepository.findOne({ email: email }, {
			populate: ['items.tenant'],
			strategy: LoadStrategy.JOINED
		});

Now I put find options into a separate object:

		const opts:FindOptions<UserEntity> = {
			populate: ['items.tenant'],
			strategy: LoadStrategy.JOINED
		}

This will result the following error message: **Type 'string' is not assignable to type 'never'.**
What am I doing wrong, how should the type look like? Thanks.

**Comment by B4nan** [comment]
You can't type that object just to `FindOptions<UserEntity>` as that translates to "no population at all" and will basically disable the type inference, either provide both type arguments or none (but that won't help you here). 

For your example, this should work:

	const opts:FindOptions<UserEntity, 'items.tenant'> = {
		populate: ['items.tenant'],
		strategy: LoadStrategy.JOINED
	}

Alternatively don't use `FindOptions` and use const assertion on the populate array:

	const opts = {
		populate: ['items.tenant'] as const,
		strategy: LoadStrategy.JOINED
	}

Note that you will never get any intellisense this way, the `FindOptions` type is designed to be used directly in the `em.find` and others, not for a separate options definition. TS unfortunately does not support partial type inference, which would be needed here.

Maybe we could add some "options builder" helper, to get around this limitation. I can imagine something like this:

```ts
const opts = createOptions(UserEntity, { ... });
```

It would basically offer you same experience as with `em.find()`.

**Reply by B4nan** [reply]
Something like this could work:

```ts
import { EntityName, FindOptions } from '@mikro-orm/core';

function createOptions<T, P extends string = never>(
	entityName: EntityName<T>,
	options: FindOptions<T, P>,
): FindOptions<T, P> {
  return options;
}

const options = createOptions(UserEntity, {
	populate: ['items.tenant'],
	strategy: LoadStrategy.JOINED
});
const user = await this.userRepository.findOne({ email: email }, options);
```

**Reply by pphunor72** [reply]
Thanks for the suggestions, both **FindOptions<UserEntity, 'items.tenant'>** and **createOptions** are suitable for me.
